[libunwind] Remove unnecessary dependencies on fprintf and stdio.h for increased baremetal friendliness

michael-kenzel opened the pull request:
libunwind already puts logging facilities behind macros so that they can be turned on and off. The two places where this isn't done yet are for debug output tracing during DWARF evaluation and compact unwinding. The code there directly calls `fprintf`. This becomes a problem when one wants to build libunwind for a baremetal target where `fprintf` isn't available. This change introduces a set of `_LIBUNWIND_TRACE_DWARF_EVAL` and `_LIBUNWIND_TRACE_COMPACT_UNWIND` macros in the style of the already existing macros to remove the direct use of `fprintf`. It also removes unnecessary includes of `stdio.h`.

michael-kenzel committed:
[libunwind] Remove unnecessary dependencies on stdio.h for increased baremetal friendliness

arichardson commented on libunwind/src/config.h:
```diff
@@ -223,6 +223,21 @@
     } while (0)
 #endif
 
+#define _LIBUNWIND_TRACING_DWARF_EVAL (0)
```
Maybe it would be best to follow the same pattern as above and allow turning it on for debug builds? But seeing as it's always been hardcoded to false, maybe the current preprocessor define makes more sense. It should be very rare that this kind of tracing is needed and a debugger might be better.

michael-kenzel replied:
Yeah, I figured if anyone ever actually needed to turn this on for debug builds in general, they'd have added that option. It seemed to me like this is more just some convenience thing for whoever was working on that code. I didn't want to couple this to `NDEBUG` or some external function like the other stuff does since this output so far wasn't turned on that way either. And adding a whole CMake build option just for this seemed like overkill. Given the way it's been used so far, doing it via a CMake option would likely get in the way more than anything else since that would force a wider rebuild every time you switch. So I just kept the overall behavior as-is. With the macro in the config header now, it will be trivial to add a build flag or something, should the need ever arise.

michael-kenzel replied:
In fact, since this is more some kind of internal feature, maybe it shouldn't be in the config header at all but rather just defined and undef'ed again inside the original function? Though having all the stuff to do with logging defined in a similar fashion in one place is probably preferable…

arichardson reviewed (approved):
I'm happy with this as is but I'd wait for at least one other positive review before merging.

ldionne commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -381,24 +381,22 @@ typename A::pint_t
-      for (pint_t *t = sp; t > stack; --t) {
-        fprintf(stderr, "sp[] = 0x%" PRIx64 "\n", (uint64_t)(*t));
-      }
+#if _LIBUNWIND_TRACING_DWARF_EVAL
```
Not attached to this line: do we want to remove the `<stdio.h>` include in this file then? But we'd need to include it from `config.h`?

michael-kenzel replied:
good question. I think it might be a good idea to completely remove `#include <stdio.h>` everywhere unless some logging is active. I can try to add that. Though maybe this should be a separate PR since I think if we do this, we should do this in the library as a whole? It seems to be included in a majority of the files. Though all those includes are kinda redundant since they all include `"config.h"` anyways, which already includes `<stdio.h>`. I guess we could just conditionally include `<stdio.h>` in `"config.h"` only when logging is active and remove the include everywhere else?

ldionne replied:
Yeah, I think that sounds like the best approach to me (i.e. removing `<stdio.h>` everywhere except in `"config.h"`). You can do it in this patch or in a follow-up, I don't mind.

michael-kenzel replied:
Actually, there seems to be more of the same: https://github.com/llvm/llvm-project/blob/d97981c98a70ebeaa8901f34921b0a69a068ff5d/libunwind/src/UnwindCursor.hpp#L1714

I just overlooked that since that path wasn't being built in my setup. I'll update this patch to do the same thing for this code too and remove the `<stdio.h>` includes.

ldionne replied:
We normally don't output like that from the tests, but I am not aware of any mechanism available from libunwind to conditionally output information. FWIW I think it would be reasonable for your patch not to change that test (even though it would be a barrier for porting the tests to bare metal).

michael-kenzel replied:
makes sense, I'll just leave the test as-is.

ldionne reviewed (approved):
This looks like an improvement to me. I have a comment/question about the `stdio.h` include which is now dangling, but apart from that I'm happy with the patch.

michael-kenzel commented:
As discussed, I've removed unnecessary includes of `<stdio.h>` as well as a couple other unnecessary includes. I've put the include of `<stdio.h>` in `config.h` under an `#ifdef` so it only gets included when logging is potentially active. I've also moved the include of `<assert.h>` from `config.h` into only the files that actually need it. In two places, I replaced the include `<stdio.h>` with an include of the header that was actually needed (`<stdint.h>` or `<string.h>`).

[AddressSpace.hpp](https://github.com/llvm/llvm-project/pull/72040/files#diff-dc10e493a84f7de241e331f66a5af195689f08d45a6fd0da90b275a662856327) was using `snprintf` to copy strings into a buffer. I replaced those uses with `strncpy` to remove the need for `<stdio.h>`.

Finally, I've added `_LIBUNWIND_TRACE_COMPACT_UNWIND` to also deal with the debug logging in [UnwindCursor.hpp](https://github.com/llvm/llvm-project/pull/72040/files#diff-e59aa90030d7262678863315066b19c20f770a3b02ef960961d7d01fa4b15596).

Unfortunately, I can't really exercise all these code paths here as these changes now span code specific to multiple different platforms. I'll trust that if all the tests pass, that means it works…

michael-kenzel commented:
As discussed, I've removed unnecessary includes of `<stdio.h>` as well as a couple other unnecessary includes. I've put the include of `<stdio.h>` in `config.h` under an `#ifdef` so it only gets included when logging is potentially active. I've also moved the include of `<assert.h>` from `config.h` into only the files that actually need it. In two places, I replaced the include `<stdio.h>` with an include of the header that was actually needed (`<stdint.h>` or `<string.h>`).

[AddressSpace.hpp](https://github.com/llvm/llvm-project/pull/72040/files#diff-dc10e493a84f7de241e331f66a5af195689f08d45a6fd0da90b275a662856327) was using `snprintf` to copy strings into a buffer. I replaced those uses with `strncpy` to remove the need for `<stdio.h>`.

Finally, I've added `_LIBUNWIND_TRACE_COMPACT_UNWIND` to also deal with the debug logging in [UnwindCursor.hpp](https://github.com/llvm/llvm-project/pull/72040/files#diff-e59aa90030d7262678863315066b19c20f770a3b02ef960961d7d01fa4b15596).

Unfortunately, I can't really exercise all these code paths here as these changes now span code specific to multiple different platforms. I'll trust that if all the tests pass, that means it works…

arichardson commented on libunwind/src/AddressSpace.hpp:
```diff
@@ -672,7 +671,7 @@ inline bool LocalAddressSpace::findFunctionName(pint_t addr, char *buf,
   if (dladdr((void *)addr, &dyldInfo)) {
     if (dyldInfo.dli_sname != NULL) {
-      snprintf(buf, bufLen, "%s", dyldInfo.dli_sname);
+      strncpy(buf, dyldInfo.dli_sname, bufLen);
```
This will always write up to 512 '\0' bytes. maybe best to use memcpy+strlen instead?

michael-kenzel replied:
Indeed. I did a quick test on my machine here:
```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
copy_snprintf         4005109 ns      4005097 ns          175
copy_strncpy          3580506 ns      3580409 ns          194
copy_memcpy_strlen    3346781 ns      3346773 ns          213
```
so doing `strlen` + `memcpy` seems to be the best option. I'll update the PR.

arichardson commented on libunwind/src/AddressSpace.hpp:
```diff
@@ -681,7 +680,7 @@ inline bool LocalAddressSpace::findFunctionName(pint_t addr, char *buf,
   char *funcName = getFuncNameFromTBTable(addr, nameLen, offset);
   if (funcName != NULL) {
-    snprintf(buf, bufLen, "%.*s", nameLen, funcName);
+    strncpy(buf, funcName, nameLen < bufLen ? nameLen : bufLen);
```
This will always write up to 512 '\0' bytes. maybe best to use memcpy+strlen instead?

michael-kenzel commented:
> [AddressSpace.hpp](https://github.com/llvm/llvm-project/pull/72040/files#diff-dc10e493a84f7de241e331f66a5af195689f08d45a6fd0da90b275a662856327) was using `snprintf` to copy strings into a buffer. I replaced those uses with `strncpy` to remove the need for `<stdio.h>`.

As discussed with @arichardson, I've replaced `strncpy` with `strlen` + `memcpy`.

arichardson reviewed (approved):
Thanks! LGTM once CI is happy.

ldionne commented:
The CI issues are flakes (we just made major changes to our infrastructure and we're still adjusting). I think this is good to go.

TiborGY commented:
@michael-kenzel Was there a reason why this never got merged?

arichardson commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -381,24 +381,22 @@ typename A::pint_t
-      for (pint_t *t = sp; t > stack; --t) {
-        fprintf(stderr, "sp[] = 0x%" PRIx64 "\n", (uint64_t)(*t));
-      }
+#if _LIBUNWIND_TRACING_DWARF_EVAL
```
Is this needed? Can't the compiler elide the loop if _LIBUNWIND_TRACE_DWARF_EVAL is a no-op?

michael-kenzel replied:
I would expect the compiler to optimize the loop away in an optimized build. If I remember correctly, and it's been a while, then the reason I did it this way was just to avoid generating a pointless loop also in an unoptimized build.

arichardson commented on libunwind/src/config.h:
```diff
@@ -223,6 +221,41 @@
+#define _LIBUNWIND_TRACE_COMPACT_UNWIND0(msg)
+#define _LIBUNWIND_TRACE_COMPACT_UNWIND(msg, ...)
+#else
+#define _LIBUNWIND_TRACE_COMPACT_UNWIND0(msg)                                  \
+  do {                                                                         \
+    fprintf(stderr, msg);                                                      \
+  } while (0)
+#define _LIBUNWIND_TRACE_COMPACT_UNWIND(msg, ...)                              \
+  do {                                                                         \
+    fprintf(stderr, msg, __VA_ARGS__);                                         \
+  } while (0)
```
```suggestion
#define _LIBUNWIND_TRACE_COMPACT_UNWIND(...)                                   \
  do {                                                                         \
    fprintf(stderr, __VA_ARGS__);                                         \
  } while (0)
```
Does this work?

michael-kenzel replied:
It probably would; the reason why I did it the way I did was consistency with the existing `_LIBUNWIND_LOG`/`_LIBUNWIND_LOG0` macros. I guess best would be to just rewrite those in your suggested way as well. But maybe this should be done in a separate PR to keep this one focused on just replacing `fprintf`?

arichardson replied:
Sure fine by me :)

arichardson commented on libunwind/src/config.h:
```diff
@@ -223,6 +221,41 @@
+#define _LIBUNWIND_TRACE_DWARF_EVAL0(msg)
+#define _LIBUNWIND_TRACE_DWARF_EVAL(msg, ...)
+#else
+#define _LIBUNWIND_TRACE_DWARF_EVAL0(msg)                                      \
```
Same as above?

michael-kenzel replied:
same as above 🙂‍↕️

arichardson reviewed (commented):
This looks like a nice improvement. Some minor questions inline

michael-kenzel commented:
> @michael-kenzel Was there a reason why this never got merged?

I don't think there was, I guess this just fell through the cracks…

michael-kenzel commented:
@arichardson as promised, I have rebased the changes. I'm not sure about the test failures, it would seem to me that they are unlikely to be caused by anything I did?

arichardson commented:
> @arichardson as promised, I have rebased the changes. I'm not sure about the test failures, it would seem to me that they are unlikely to be caused by anything I did?

Yes it looks like an issue with the android CI setup, nothing related to this patch I believe.
